src/SendGrid/Permissions/SendGridPermissionsBuilder.cs

```diff
@@ -0,0 +1,115 @@
+            }
+
+            return scopes.Distinct().ToArray();
+        }
+
+        /// <summary>
+        /// Adds the permissions for the specified <typeparamref name="TScope"/>
+        /// </summary>
+        /// <typeparam name="TScope">The type of the scope.</typeparam>
+        /// <param name="options">The o.</param>
+        /// <returns>The builder instance with the permissions added.</returns>
+        public SendGridPermissionsBuilder AddPermissionsFor<TScope>(ScopeOptions options)
+            where TScope : ISendGridPermissionScope, new()
+        {
+            var scope = new TScope();
+
+            if ((scope.IsMutuallyExclusive && this.addedPermissions.Any()) || this.addedPermissions.Any(x => x.Key.IsMutuallyExclusive))
+            {
+                throw new InvalidOperationException($"{scope.Name} permissions are mutually exclusive from all others. An API Key can either have {scope.Name} Permissions, or any other set of Permissions.");
+            }
+
+            this.addedPermissions[scope] = options;
+            return this;
+        }
+
+        /// <summary>
+        /// Adds the permissions for the specified <typeparamref name="TScope"/> with all possible scopes requested
+        /// </summary>
+        /// <typeparam name="TScope">The type of the scope.</typeparam>
+        /// <returns>The builder instance with the permissions added.</returns>
+        public SendGridPermissionsBuilder AddPermissionsFor<TScope>()
+            where TScope : ISendGridPermissionScope, new()
+        {
+            return this.AddPermissionsFor<TScope>(ScopeOptions.All);
+        }
+
+        /// <summary>
+        /// Filters the emitted scopes based on the scopes available to the current API key.
+        /// </summary>
+        /// <param name="client">The client.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The builder instance with the filter applied.</returns>
+        public async Task<SendGridPermissionsBuilder> FilterByCurrentApiKeyAsync(ISendGridClient client, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var response = await client.RequestAsync(SendGridClient.Method.GET, null, null, "scopes", cancellationToken);
+            var body = await response.DeserializeResponseBodyAsync(response.Body);
+            var userScopesJArray = (body["scopes"] as JArray);
+            var includedScopes = userScopesJArray.Values<string>().ToArray();
+            this.Exclude(scope => !includedScopes.Contains(scope));
+            return this;
+        }
+
+        /// <summary>
+        /// Adds an exclusion filter the will not emit any scopes that the filter matches.
+        /// </summary>
+        /// <param name="filter">The filter.</param>
+        /// <returns>The builder instance with the exclusion filter applied.</returns>
+        public SendGridPermissionsBuilder Exclude(Func<string, bool> filter)
```

Comment:
Would also like to see an `Include` function, ideally one that also accepts a list of scopes, so it can easily be built from the returned body of the GET /scopes api call

Reply:
Should such a method mask against a list of known scopes so that arbitrary (e.g. invalid) scopes cannot be included? I believe the request will ultimately fail if it is called with scopes that don't actually exist. If so, should the `Build` method throw if invalid scopes have been included or should they just be omitted from the final list of omitted scopes?

Reply:
The function should accept based on a list of known scopes. If the function is called with an invalid scope, that doesn't exist, the method should throw.